text/0000-modifiers.md

```diff
@@ -0,0 +1,323 @@
+```hbs
+<input type="date" {{datepicker changeMonth=true changeYear=true}} />
+```
+
+```js
+import Modifier from '@ember/modifier';
+import $ from 'jquery';
+
+export Modifier.extend({
+  didInsertElement(_params, hash) {
+    $(this.element).datepicker(this._normalizeOptions(hash));
+  },
+
+  _normalizeOptions(options) {
+    return Object.assign(options, { minDate: 20, maxDate: '+1M +10D' });
+  },
+
+  willDestroyElement() {
+    $(this.element).datepicker('destroy');
+  }
+});
+```
+
+### Page View Tracking
+```hbs
+<section {{track-impression eventCategory="Post"}}>
+  <header>Chad liked a post</header>
+  <img src="cat.jpg">
+  {{!-- Snip --}}
+<section>
+```
+
+```js
+import Modifier from '@ember/modifier';
+import { inject as service } from '@ember/service';
+
+export Modifier.extend({
+  ga: service('google-analytics'),
+  init() {
+    this._super(...arguments);
+    this.eventCategory = undefined;
+    this.interSectionObserver = new IntersectionObserver(entries => {
+      entries.forEach(entry => {
+        this.ga.send('event', 'impression', this.eventCategory);
+      });
+    });
+  },
+  didInsertElement(_params, { eventCategory }) {
+    this.eventCategory = eventCategory;
+    this.interSectionObserver.observe(this.element);
+  },
+
+  willDestroyElement() {
+    this.interSectionObserver.unobserve(this.element);
+  }
+});
+```
+
+## How we teach this
+
+While user defined element modifiers are a new concept, Ember has had the [`{{action}}` modifier](https://www.emberjs.com/api/ember/release/classes/Ember.Templates.helpers/methods/action?anchor=action) since 1.0.0. We would need to update the existing documentation around `{{action}}` to refer to it as a modifier. In terms of documenting the modifier base class, I think we can largely repurpose existing documentation around [component's `didInsertElement`](https://guides.emberjs.com/release/components/the-component-lifecycle/#toc_integrating-with-third-party-libraries-with-didinsertelement) hook.
+
+In terms of guides, I believe we should add a section to the "Templates" section to outline how to write modifiers. This would be similar to the ["Writing Helpers"](https://guides.emberjs.com/release/templates/writing-helpers/) guide.
+
+## Drawbacks
+
+The drawbacks of adding element modifiers largely deal with explaining when to use a classic component for encapsulating some DOM manipulation versus when to use an element modifier. That being said it is likely that the recommendation would be to use element modifiers if you are just manually modifying the DOM.
+
+This API also doesn't attempt to create a corollary of `this.element` for Glimmer Components and instead offers a different API for altering DOM nodes directly. This expands the surface area of Ember's API.
+
+## Alternatives
+
+The alternative to this is to create a "ref"-like API that is available in [other client side frameworks](https://reactjs.org/docs/refs-and-the-dom.html). This may look like the following:
```

Comment:
I think that modifiers (as proposed here) allows a "ref-like api" to be implemented. Slightly modified example:

```hbs
<section>
  <h1 {{ref this "heading"}}>Hello!</h1>
  <p>How are you?</p>
</section>
```

Could easily be implemented as a modifier:

```js
export default class RefModifier extends Modifier {
  didInsertElement([component, name]) {
    component[name](this.element);
  }

  willDestroyElement([component, name]) {
    component[name](null);
  }
}
```

In other words, I don't think "refs" is an alternative to modifiers but instead is a valid use case of them...